### PR TITLE
ECMWF API Adapter and Weather Provider

### PR DESCRIPTION
**Pull request by dunnlu**
Added ecmwf_adapter, which calls the open meteo API through the ECMWF model, (with Open Meteo's python code for requests). 

Added weather_provider_ecmwf, which calls get_current and get_historical from ecmwf_adapter, parses through the data, creates a weather datum object, and adds the data to the weather datum, returning a list of weather datums correlating with the requested list of locations. 

Added demo_weather_provider_ecmwf, which demos the basic functionalities of the new weather provider.

Added test_ecmwf_adapter, which tests the basic functionalities of the ecmwf_adapter, in this case testing whether the requested location and returned location are the same..

Added test_weather_provider_ecmwf, which tests the instantiation and data storage of the weather provider. For now, it calls for actual data, but in the near future, I will create a fake WeatherApiResponse and api adapter so that the testing does not actually call for any data, but instead just tests the functionality of the weather provider itself. 

Adjusted the conda environment and environment_gpu to include the new dependencies, for compatibility with the new Open Meteo python API call.

### CODE REVIEW

**Comment by orion-junkins** [comment]
For failing tests, a lot of them were caused by some left over work in progress Code in the old api weather provider. I went ahead and reverted that and comitted which fixed a lot of them. https://github.com/orion-junkins/river-level-forecasting/pull/184/commits/f49330aad56524f18d935195549cf431f858c983

Shouldn't have affected your adapter but we can revert if we need to. 

Looks like there are still two failures related to ECMWF code that are identical to what I was getting trying to run the demo script. These are the main thing you need to fix. Is demo script working on your end? 

Also I get an error regarding incorrect datetime indexing but this also appears in main - I'll see if I can fix this one tonight.

Also looks like a bunch of small flake 8 issues - VScode has python formatting built in which makes it faster to fix (I think shift + Ctrl + P and search "format" should get you there).

Also some mypy issues, let me know if mypy passes local but git still has issues, sometime some weirdness in remote environment with mypy.

Overall code looks nice!! Just some things to iron out and we can merge 👍

**Comment by dunnlu** [comment]
All tests now pass. The only caveat is that mypy skips a lot of the checks on the new packages installed (included in environment.yml at the bottom), since they do not have associated library stubs.

File: scripts/upload/weather/historical_ecmwf_meteo_to_aws.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
+# Helper script for uploading historical weather from OpenMeteo to AWS. Intended to be modified & run as needed to upload new historical datasets.
```
Comment (orion-junkins) [inline scripts/upload/weather/historical_ecmwf_meteo_to_aws.py:1]: Not really a great automated way to do it, but I want to stress test these scripts pretty aggressively before we rely on them for too much. Fine to merge, no action items, just something to keep in mind that we can talk about next meeting.

File: src/rlf/forecasting/data_fetching_utilities/weather_provider/open_meteo/ecmwf_adapter.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+            "hourly": hourly_params,
+            "timezone": "GMT",
+            "models": "ecmwf_ifs"
+        }
+
+        return openmeteo.weather_api(url, params=params)[0]
+
+    def get_current(self,
+                    coordinate: Coordinate,
+                    past_days: int = 92,
```
Comment (orion-junkins) [inline src/rlf/forecasting/data_fetching_utilities/weather_provider/open_meteo/ecmwf_adapter.py:65]: These two args are unused

File: src/rlf/forecasting/data_fetching_utilities/weather_provider/open_meteo/ecmwf_adapter.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+        hourly_params = columns if columns is not None else self.forecast_hourly_parameters
+
+        url = "https://api.open-meteo.com/v1/ecmwf"
+        params = {
+            "latitude": coordinate.lat,
+            "longitude": coordinate.lon,
+            "hourly": hourly_params,
+            "past_days": 92,
+            "forecast_days": 7
```
Comment (orion-junkins) [inline src/rlf/forecasting/data_fetching_utilities/weather_provider/open_meteo/ecmwf_adapter.py:92]: Use them here. Dont hardcode the 92, 7, make them default parameters above and use those variables. Allows us to pass something else in if desired.

File: src/rlf/forecasting/data_fetching_utilities/weather_provider/open_meteo/parameters.py [hunk 4]
```diff
@@ -1,77 +1,159 @@
...
+        "vapour_pressure_deficit",
+        "wind_speed_10m",
+        "wind_direction_10m",
+        "wind_gusts_10m",
+        "surface_temperature",
+        "soil_temperature_0_to_7cm",
+        "runoff",
+        "total_column_integrated_water_vapour"
+    ],
+    "ecmwf_shared": [
```
Comment (orion-junkins) [inline src/rlf/forecasting/data_fetching_utilities/weather_provider/open_meteo/parameters.py:132]: Basically go through this list, and for every variable, load the data in the web browser with past_days = 92 (3 months). If the data does not cover the full 92 days, delete it. See this example:
https://open-meteo.com/en/docs/ecmwf-api#hourly=temperature_2m,relative_humidity_2m&past_days=92
Temp_2m has data for the whole window but relative humidity does not, so drop humidity. Just keep the ones that have coverage for the full three months for now. 

We can circle back to this later if we decide we want some of those features.

File: tests/forecasting/data_fetching_utilities/weather_provider/open_meteo/test_ecmwf_adapter.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+
+@pytest.mark.slow
+def test_get_current_columns_subset(fake_coordinate):
+    adapter = OpenMeteoECMWFAdapter()
+    response = adapter.get_current(coordinate=fake_coordinate, past_days=1, forecast_days=1, columns=REAL_COLUMNS)
+
+    assert pytest.approx(REAL_LONGITUDE, rel=1e-2) == response.Longitude()
+    assert pytest.approx(REAL_LATITUDE, rel=1e-2) == response.Latitude()
+    assert response.Hourly().VariablesLength() == len(REAL_COLUMNS)
```
Comment (orion-junkins) [inline tests/forecasting/data_fetching_utilities/weather_provider/open_meteo/test_ecmwf_adapter.py:58]: Nice, I'd say this is plenty for the adapter for now at least. Good work.

File: tests/forecasting/data_fetching_utilities/weather_provider/test_api_weather_provider_ecmwf.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+
+
+
+Goal for initial test:
+
+Create an intiitialization of the weather provider
+
+'''
+
```
Comment (orion-junkins) [inline tests/forecasting/data_fetching_utilities/weather_provider/test_api_weather_provider_ecmwf.py:29]: Tests look good but get rid of this extra comment.

**Review by orion-junkins — CHANGES_REQUESTED** [review]
Really nice work! I made a couple of small tweaks myself (see recent commits, main change was figuring out how to get a non empty units dict) and highlighted a couple for you to work on. 

1. Prune the shared param list (looks like some variables don't have complete recent data so exclude those for now)
2. Update the demo script to only use valid variables after that pruning.
3. Remove extraneous comments in one or two places I highlighted

Feel free to make a single commit with something like "small tweaks following PR review". Resolve my comments as you address each.

So long as everything is still green we will be ready to merge once those last couple of changes are made :)

Congrats on the progress, super nice job.

**Pull request merged**